api/src/main/java/jakarta/data/repository/KeysetPageable.java

```diff
@@ -0,0 +1,265 @@
+ *
+ * <pre>
+ * Employee emp = ...
+ * KeysetPageable pagination = Pageable.size(50).afterKeyset(emp.lastName, emp.firstName, emp.id);
+ * page = employees.findByHoursWorkedGreaterThan(1500, pagination);
+ * </pre>
+ *
+ * <p>By making the query for the next page relative to observed values,
+ * not a numerical position, keyset pagination is less vulnerable to changes
+ * that are made to data in between page requests. Adding or removing entities
+ * is possible without causing unexpected missed or duplicate results.
+ * Keyset pagination does not prevent misses and duplicates if the entity
+ * properties which are the sort criteria for existing entities are modified
+ * or if an entity is re-added with different sort criteria after having
+ * previously been removed.</p>
+ *
+ * <h2>Keyset Pagination with &#64;Query</h2>
+ *
+ * <p>Keyset pagination involves generating and appending to the query
+ * additional query conditions for the keyset properties. In order for that to
+ * be possible, the user-provided JPQL query must end with a
+ * <code>WHERE</code> clause to which additional conditions can be appended.
+ * Enclose the entire conditional expression of the <code>WHERE</code> clause
+ * in parenthesis.
+ * Sort criteria must be specified independently from the user-provided query,
+ * either with the {@link OrderBy} annotation or
+ * <code>Pageable</code> {@link Pageable#of(long, long, Sort...) Sort} parameters.
+ * For example,</p>
+ *
+ * <pre>
+ * &#64;Query("SELECT o FROM Customer o WHERE (o.ordersPlaced &gt;= ?1 OR o.totalSpent &gt;= ?2)")
+ * &#64;OrderBy("zipcode")
+ * &#64;OrderBy("birthYear")
+ * &#64;OrderBy("id")
+ * KeysetAwareSlice&lt;Customer&gt; getTopBuyers(int minOrders, float minSpent, Pageable pagination);
+ * </pre>
+ *
+ * <h2>Page Numbers and Totals</h2>
+ *
+ * <p>Page numbers, total numbers of elements across all pages, and total
+ * count of pages are not accurate when keyset pagination is used and should
+ * not be relied upon.</p>
+ */
+public class KeysetPageable extends Pageable {
+    /**
+     * Direction of keyset pagination.
+     */
+    public static enum Mode {
+        /**
+         * Indicates forward keyset pagination, which follows the
+         * direction of the {@link OrderBy} annotations, <code>Pageable</code>
+         * {@link Pageable#of(long, long, Sort...) Sort} parameters,
+         * or <code>OrderBy</code> name pattern of the repository method.
+         */
+        NEXT,
+
+        /**
+         * Indicates reverse keyset pagination. which follows the opposite
+         * direction of the {@link OrderBy} annotations, <code>Pageable</code>
+         * {@link Pageable#of(long, long, Sort...) Sort} parameters,
+         * or <code>OrderBy</code> name pattern of the repository method.
+         */
+        PREVIOUS
+    }
+
+    /**
+     * Represents keyset values, which can be a starting point for
+     * requesting a next or previous page.
+     */
+    public static class Cursor {
```

Comment:
I think I would prefer to have an interface here and make the implementation private. If someone wants a different backing storage than an `Object[]` or want to provide some extensions to the cursor like the type of individual elements, allowing custom implementations is very useful.